### PR TITLE
[new task] Jet App Interaction

### PR DESCRIPTION
**Pull request by serhiipylypchuk1991**
https://docs.google.com/document/d/1YEqmPlyi7n05zXpzrGNq2vZA9wDwM6Jm9pedCYcVxk8/edit#heading=h.legknrh56du6 
Lesson 2. Webix Jet Interaction. Practice

### CODE REVIEW

File: sources/views/contacts.js [hunk 1]
```diff
@@ -1,81 +1,89 @@
...
+		const _ = this.app.getService("locale")._;
+
+		const add_new_button = {
+			view:"button",
+			value:_("Add new"),
+			css:"webix_primary",
+			click:()=>{
+				webix.message({text:_("New element was added"), expire:350});
+				const new_contact = { Name:"Name", Email:"Email", Status:0, Country:0};
+				contacts.add(new_contact);
```
Comment (Zwillinge) [inline sources/views/contacts.js]: you need to edit the new item immediately after adding

File: sources/views/contacts.js [hunk 2]
```diff
@@ -1,81 +1,89 @@
...
+						text:_("Do you still want to continue?"),
+						type:"confirm-warning"
+					}).then(() => {
+
+						webix.message({
+							text:_("Element was deleted"),
+							type:"info"
+						});
+
+						const prev_id = this.getPrevId(id);
```
Comment (Zwillinge) [inline sources/views/contacts.js]: it is cheating. also - what if you remove the last item from the contacts?

how it should be solved: you show the url `top/contacts` and the contacts module decides in urlChange which item to select

File: sources/views/contacts.js [hunk 3]
```diff
@@ -1,81 +1,89 @@
...
-				},
-				Email:webix.rules.isEmail
+			select:true,
+			on:{
+				onAfterSelect:function(id){
+					this.$scope.setParam("id", id, true);
+				}
+			},
+			onClick:{
+				remove_list_item_btn:function(e, id){
```
Comment (Zwillinge) [inline sources/views/contacts.js]: it's better to store action handles as jetview methods

File: sources/views/contacts.js [hunk 4]
```diff
@@ -1,81 +1,89 @@
...
+
+		if(id_url && contacts.exists(id_url)){
+			list.select(id_url);
 		}else{
-			webix.message({
-				text:"Please, enter the correct data in the fields of the form",
-				type:"error",
-				expire:3000
-			});
+			list.select(list.getFirstId());
```
Comment (Zwillinge) [inline sources/views/contacts.js]: what if the list does not contain elements?

File: sources/views/contacts.js [hunk 5]
```diff
@@ -1,81 +1,89 @@
...
-			});
-			form_collection.clear();
+
+	init(view){
+		const list = view.queryView("list");
+		const id_url = this.getParam("id");
+
+		list.parse(contacts);
+
+		if(id_url && contacts.exists(id_url)){
```
Comment (Zwillinge) [inline sources/views/contacts.js]: `init` is not the best place for item selection logic. `urlChange` maybe better as we need to select the item not only on module initialization

also: in your app the item is not re-selected after page refresh (F5) as required per task

File: sources/views/grid_view.js [hunk 6]
```diff
@@ -1,47 +1,52 @@
...
+		const add_new_button = {
 			view:"button",
-			value:"Add new",
+			value:_("Add new"),
 			css:"webix_primary",
 			click:()=>{
-				webix.message({text:"New element was added", expire:350});
-				this.grid.add({Name:""});
+				webix.message({text:_("New element was added"), expire:350});
+				this.grid.add({id:0, value:"New"});
```
Comment (Zwillinge) [inline sources/views/grid_view.js]: do not use ids while adding new items (they can be non unique). just omit them - the view will generate the unique id automatically

do not use `id:0` - it is forbidden for webix views, as it means no id

File: sources/views/top.js [hunk 7]
```diff
@@ -1,30 +1,50 @@
...
-		var ui = {
+
+		const ui = {
 			cols:[
-				{rows:[{rows:[header, menu]}]},
-				{rows:[
-					{ $subview:true }
-				]}
+				{rows:[{ rows:[header, sidebar]}] },
+				{rows:[{ $subview:true }]}
```
Comment (Zwillinge) [inline sources/views/top.js]: why do you need to wrap subview into `rows` ?

**Review by Zwillinge — CHANGES_REQUESTED** [review]

File: sources/views/contacts.js (hunk 1 above)
Comment (Zwillinge) [reply sources/views/contacts.js]: `getLastId()` is not reliable as you may need to add to any other position as well. 
for client-side data it is better to use `var id = some.add()`

File: sources/views/contacts.js [hunk 8]
```diff
@@ -1,81 +1,87 @@
...
+	}
+
+	urlChange(view){
+		const list = view.queryView("list");
+		const url_id = this.getParam("id");
+		const first_id = contacts.getFirstId();
+
+		if(url_id && contacts.exists(url_id)){
+			list.select(url_id);
+		}else if(first_id){
```
Comment (Zwillinge) [inline sources/views/contacts.js]: small optimization: you do not need to define a variable (first_id) that you will probably not use 

the code may look like 
```
const id = this.getParam("id") || contacts.getFirstId();
if(contacts.exists(id))
   list.select(id);
```

File: sources/views/form.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		};
+
+		return form;
+	}
+
+	urlChange(view){
+		const url_id = this.getParam("id");
+		const exist_id = contacts.exists(url_id);
+
+		if(url_id && exist_id){
```
Comment (Zwillinge) [inline sources/views/form.js]: again small optimization `if(id) && contacts.exists(id){ .. }`

File: sources/views/grid_view.js [hunk 10]
```diff
@@ -1,67 +1,82 @@
...
+
 		return {margin:3, rows:[add_new_button,datatable]};
 	}
+
+	addNewElement(){
+		const _ = this.app.getService("locale")._;
+		
+		webix.message({text:_("New element was added"), expire:350});
+		this.data.add({value:"New"});
+		scrollToLastAddedElemen(this.grid,true,true);
```
Comment (Zwillinge) [inline sources/views/grid_view.js]: I do not really like this function, as it relies on the order (`getLastId()`)
you can pass id that is returned by the `.add()` method

File: sources/views/contacts.js [hunk 11]
```diff
@@ -1,81 +1,87 @@
...
-				text:"Please, enter the correct data in the fields of the form",
-				type:"error",
-				expire:3000
+				text:_("Element was deleted"),
+				type:"info"
 			});
+
+			contacts.remove(id);
+			this.show("contacts");
+			return false;
```
Comment (Zwillinge) [inline sources/views/contacts.js]: this `return false` is inside promise handler, so it no longer blocks item selection when users clicks on the remove icon. 
the same problem can be spotted in gridview

File: sources/views/contacts.js (hunk 8 above)
Comment (serhiipylypchuk1991) [reply sources/views/contacts.js]: It is a really cool optimization

File: sources/views/contacts.js (hunk 11 above)
Comment (serhiipylypchuk1991) [reply sources/views/contacts.js:60]: ok

File: sources/views/grid_view.js (hunk 10 above)
Comment (serhiipylypchuk1991) [reply sources/views/grid_view.js]: I took it into account

File: sources/views/form.js (hunk 9 above)
Comment (serhiipylypchuk1991) [reply sources/views/form.js]: if(id && contacts.exists(id)){ .. }
ok

File: sources/views/contacts.js (hunk 8 above)
Comment (serhiipylypchuk1991) [reply sources/views/contacts.js]: ok. it looks better

File: sources/views/grid_view.js (hunk 6 above)
Comment (serhiipylypchuk1991) [reply sources/views/grid_view.js]: ok

File: sources/views/contacts.js (hunk 5 above)
Comment (serhiipylypchuk1991) [reply sources/views/contacts.js]: ok

File: sources/views/contacts.js (hunk 3 above)
Comment (serhiipylypchuk1991) [reply sources/views/contacts.js]: ok

File: sources/views/contacts.js (hunk 11 above)
Comment (serhiipylypchuk1991) [reply sources/views/contacts.js]: ok. I removed this expression from handlers

File: sources/views/grid_view.js (hunk 10 above)
Comment (Zwillinge) [reply sources/views/grid_view.js]: just a small thing 

you pass id to the method and inside it check whether there is an id (and if there's no id, you do nothing). 
if is better to check id before method call `if(id) scrollTo..`, but it does not make sense in your case 

so you can just remove the check of id inside the method

File: sources/views/contacts.js (hunk 11 above)
Comment (Zwillinge) [reply sources/views/contacts.js]: it does not solve the problem, item must not be selected if a user clicks the remove icon 
here it should be moved out of the method
```
webix.confirm().then..;
return false;
```

File: sources/views/grid_view.js (hunk 10 above)
Comment (serhiipylypchuk1991) [reply sources/views/grid_view.js]: ok

File: sources/views/contacts.js (hunk 11 above)
Comment (serhiipylypchuk1991) [reply sources/views/contacts.js]: Do I have to remove confirmation at all?

Comment (Zwillinge) [reply sources/views/contacts.js]: no
you just place `return false` after webix.confirm call

Comment (serhiipylypchuk1991) [reply sources/views/contacts.js]: ok, it is clear now

**Pull request merged**